fixed the hasApproval implementation

grodrigues3 opened the pull request:
Now we allow for approval for any approver in path
Also allow recognize approve cancel comment

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/contrib/1923)

<!-- Reviewable:end -->

grodrigues3 committed:
fixed the hasApproval implementation

Now we allow for approval for any approver in path
Also allow recognize approve cancel comment

grodrigues3 commented:
@foxish @apelisse @hongchaodeng 
Note that I tested this implementation on #1914 and it correct *adds and *removes the label as expect. *I ran this in dry-run mode

hongchaodeng commented on mungegithub/issue-labeler/OWNERS:
```diff
@@ -1,4 +1,4 @@
-assignees
+assignees:
```
Can we put this into another commit?

grodrigues3 replied:
you mean separate PR or just a separate commit? It was mostly for testing purposes

hongchaodeng replied:
Commit

> It was mostly for testing purposes

Understand. That's why we should have separate commit.

hongchaodeng commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -108,37 +106,28 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
 	for _, file := range files {
-		if !h.hasApproval(*file.Filename, approverSet, maxDepth) {
+		fileOwners := h.features.Repos.Assignees(*file.Filename)
+		if fileOwners.Intersection(cancelSet).Len() > 0 {
```
Can you add the algorithm of cancelling into function comments too?
I think the comments doesn't reflect well enough what is being changed.

hongchaodeng reviewed (commented):
Reviewed the approver handler.
Not sure about the "repo-updates.go".

Thanks @grodrigues3 for doing the work.

grodrigues3 commented:
any other thoughts before apply lgtm label?  Anyone have reasonable ideas for automatically testing this?

grodrigues3 commented:
@apelisse can you PTAL?

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -64,15 +61,17 @@ func (*ApprovalHandler) AddFlags(cmd *cobra.Command, config *github.Config) {}
+// - Initially, we build two sets, an approverSet and a cancelSet
+//   - Go through all comments after latest commit.
+//	- If anyone said "/approve", add them to approverSet.
+//	- If anyone said "/approve cancel", add them to cancelSet.
```
So what happens if I do (in three different comments):
```
/approve
/approve cancel
/approve
```
First I'm in the `approverSet`, then I'm in `cancelSet`, and the last `/approve` is a no-op?

Suggested improvement:
- Add to `approverSet` when you see `/approve`.
- Remove from `approverSet` when you see `/approve cancel`
- Remove label if already applied and shouldn't be anymore

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -64,15 +61,17 @@ func (*ApprovalHandler) AddFlags(cmd *cobra.Command, config *github.Config) {}
+//     - in one of the file's parent directories
+//   - If a valid approver is in cancelSet, we remove the approve label (if there is one) and return immediately
+// - Iff all files have been approved, the bot will add the "approved" label.
+// - This implies that if any file doesn't have "approved" from a valid approver, the approval label will not be applied
```
Not for this PR, but we definitely need to tell people what files are not approved, and who can approve

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -103,42 +103,30 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+			cancelSet.Insert(*c.User.Login)
 		}
 	}
+	glog.Infof("This is the cancel set %v", cancelSet)
```
Please remove debug code :-)

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,6 +90,7 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
 	}
 
 	approverSet := sets.String{}
+	approverSet.Insert(*obj.Issue.User.Name)
```
BCNTC (Better check null than crash, I just made that up). Applies specifically well when the data is being returned by a third party interface.

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,6 +90,7 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+	approverSet.Insert(*obj.Issue.User.Name)
 
 	// from oldest to latest
 	for i := len(comments) - 1; i >= 0; i-- {
```
Can you please maybe break it into function for readability: `CreateApproverSet() sets.String` here instead of the loop

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -103,42 +104,32 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
 
+	//Checks that no valid approver has commented to cancel the approve label since last commit
+	//Checks that all files have an approver in the approverSet
 	for _, file := range files {
```
New function: `AreAllFilesApproved() bool`.

(You're mixing the approval checking logic with the action that it implies)

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -103,42 +104,32 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
 	}
+
+	//every file has been approved by a valid approver
 	obj.AddLabel(approvedLabel)
```
Doesn't have to be in a function: if approved addLabel, else RemoveLabel.

Note that my patch needs to go in first if you don't want to check for the presence of the label before you add/remove.

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -103,42 +104,39 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+			if obj.HasLabel(approvedLabel) {
+				obj.RemoveLabel(approvedLabel)
+			}
 			return
```
Early return is wrong now

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,8 +90,31 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
 
 	approverSet := sets.String{}
+	if obj.Issue.User.Name != nil {
+		approverSet.Insert(*obj.Issue.User.Name)
```
@fejta Automatically adding author to the `approverSet` should be the simplest way to make the process simple when you modify your own code/area.

apelisse reviewed (approved):
Awesome! To be honest, even though you made the logic more complicated than before, it's ten times easier to understand what it does ...

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,8 +90,31 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+		approverSet.Insert(*obj.Issue.User.Name)
+	}
 
 	// from oldest to latest
```
Dangling comment

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,8 +90,31 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
 	// from oldest to latest
+	approverSet = approverSet.Union(createApproverSet(comments))
+	//Checks that no valid approver has commented to cancel the approve label since last commit
+	//Checks that all files have an approver in the approverSet
```
Dangling comments

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,8 +90,31 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+	//Checks that all files have an approver in the approverSet
+	needsApproval := h.getApprovalNeededFiles(files, approverSet)
+
+	//every file has been approved by a valid approver
```
Comment says the opposite (maybe just remove it)

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,8 +90,31 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+			obj.RemoveLabel(approvedLabel)
+		}
+	} else {
+		obj.AddLabel(approvedLabel)
```
Above you test if the label is already there, here you don't check if it's already there or not. You probably should

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -64,15 +62,17 @@ func (*ApprovalHandler) AddFlags(cmd *cobra.Command, config *github.Config) {}
-// - Initially, we set up approverSet
-//   - Go through all comments after latest commit. If any approver said "/approve", add him to approverSet.
-// - For each file, we see if any approver of this file is in approverSet.
+// - Initially, we build two sets, an approverSet and a cancelSet
```
Comments needs an update

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -21,14 +21,21 @@ import (
 )
 
+const (
+	approvalNotificationName = "ApprovalNotifier"
```
Not super fan of the name, maybe `[APPROVED-FILES]` ? But let's not bikeshed too much...

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -109,6 +118,60 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+			approvedFiles.Insert(*fn.Filename)
+		}
+	}
+	context := bytes.Buffer{}
```
https://golang.org/pkg/bytes/#NewBufferString

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -109,6 +118,60 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+		}
+	}
+	context := bytes.Buffer{}
+	context.WriteString("The following files have been approved:\n")
```
Not super fan of splitting the two lists

apelisse replied:
But whatever :-)

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -109,6 +118,60 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+	context.WriteString("The following files have been approved:\n")
+	fileFmt := "\t%s"
+	for _, fn := range approvedFiles {
+		context.WriteString(fmt.Sprintf(fileFmt, fn))
```
should be the `OWNERS` file, not the file per se. Or we will have too many of these. And it'd be nice if it's a link. and if we can see who approved what ;-)

grodrigues3 replied:
you never explained how i get the OWNERs file

apelisse replied:
I don't know either. One needs to think about it

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,8 +96,88 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+	needsApproval := h.getApprovalNeededFiles(files, approverSet)
+
+	if needsApproval.Len() > 0 {
+		glog.Infof("Canceling the approve label because these files %v have no valid approvers", needsApproval)
```
This log doesn't seem right. It's not cancelling anything.

grodrigues3 replied:
3 lines below, it removes the "approved" label.  Isn't that a cancel?

apelisse replied:
=))

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,8 +96,88 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+	if needsApproval.Len() > 0 {
+		glog.Infof("Canceling the approve label because these files %v have no valid approvers", needsApproval)
+		//if latest notification stale, update it
+		if notificationNeedsUpdate(obj) {
```
By reading this code, I don't know if it removes it or not. Maybe rename the function?

grodrigues3 replied:
can you suggest a newName? deleteOldNotificationsAndCheckForUpdate seems a little long?

apelisse replied:
It probably does too much then. `findObsoleteNotification() *IssueComment`. And remove it outside the function.

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,8 +96,88 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
 
-	// from oldest to latest
+func createMessage(obj *github.MungeObject, filesNeedApproval sets.String) {
+	files, err := obj.ListFiles()
```
You could just pass-in files directly, you have it in the parent function

grodrigues3 replied:
sure, but I need the obj for Posting the notification below regardless

apelisse replied:
the name is wrong anyway. Can you return the message and post it from the calling function?

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,8 +96,88 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+	} else if !obj.HasLabel(approvedLabel) {
+		obj.AddLabel(approvedLabel)
+	}
+
```
extra-space :-)

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,8 +97,117 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+	}
+}
+
+func findObscoleteNotification(obj *github.MungeObject) *goGithub.IssueComment {
```
minor nit: obsolete is spelled obsolete

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,8 +97,117 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+	return nil
+
+}
+func notificationNeedsUpdate(obj *github.MungeObject) bool {
```
duplicate?

Commit:
mungegithub: Update repo-updates to read reviewers

Actually this commit does (unfortunately) much more:
Update repo-updates to:
- Read reviewers
- And update blunderbuss to use reviewers rather than assignees
- Read assignees and approvers into approvers rather than assignees
- And update approval-handler to use approvers instead
- removes comments about owners

Commit:
mungegithub: top-level approvers can approve everything

Top-level approvers can approve everything below their directory, not
only the directory where the file is. Use `Approvers()` rather than
`LeafApprovers()` to reflect that.

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -91,8 +97,90 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+	fileFmt := "- %s\n"
+	for _, fn := range approvedList {
+		context.WriteString(fmt.Sprintf(fileFmt, fn))
+	}
```
Append extra `\n` after this list

grodrigues3 committed:
fix

grodrigues3 committed:
notify approvers that they must approve

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,62 +95,161 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
 		}
 	}
-	obj.AddLabel(approvedLabel)
+	context := bytes.NewBufferString("The PR requires approval from at least one person in the following OWNERS files:\n")
```
Oh so you removed the list of already approved files?

grodrigues3 replied:
yeah, just shows the OWNERS files that need an approver to sign off

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,62 +95,161 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
-		if !h.hasApproval(*file.Filename, approverSet, maxDepth) {
-			return
+		// find owners files for files that have not yet been approved
+		if !filesNeedApproval.Has(*file.Filename) {
```
You should do this directly in `getApprovalNeededFiles`

apelisse replied:
So if a file is NOT in `filesNeedApproval`, we looks for it `OWNERS` path? that seems wrong.

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,62 +95,161 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+}
+
+// isApproved indicates whether or not someone from the list of OWNERS for a file has approved the PR
+func (h ApprovalHandler) isApproved(someFile *goGithub.CommitFile, approverSet sets.String) bool {
```
This could operate directly on the list of OWNERS file instead, but whatever

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,62 +95,161 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+	needsApproval := sets.String{}
+	for _, file := range files {
+		if !h.isApproved(file, approverSet) {
+			needsApproval.Insert(*file.Filename)
```
This should probably generate the OWNERS file path directly

grodrigues3 commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -17,20 +17,24 @@ limitations under the License.
+	"fmt"
 	"github.com/golang/glog"
-	githubapi "github.com/google/go-github/github"
+	goGithub "github.com/google/go-github/github"
```
deleting your comment

grodrigues3 commented on mungegithub/features/repo-updates.go:
```diff
@@ -184,15 +172,15 @@ func (o *RepoInfo) updateRepoUsers() error {
 
-	o.assignees = map[string]sets.String{}
-	//o.owners = map[string]sets.String{}
+	o.approvers = map[string]sets.String{}
```
should be a make here?  And/or set.NewString{}

apelisse replied:
Nah, I think this is fine

grodrigues3 commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,62 +95,157 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+// them to the approverSet.  The function uses the latest approve or cancel comment
+// to determine the Users intention
+func createApproverSet(comments []*goGithub.IssueComment) sets.String {
+	approverSet := sets.String{}
```
sets.NewString()

grodrigues3 commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,62 +95,157 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+// getApprovalNeededFiles identifies the files that still need approval from someone in their OWNERS files
+func (h ApprovalHandler) getApprovalNeededOwnersFiles(files []*goGithub.CommitFile, approverSet sets.String) (sets.String, sets.String) {
+	owners := sets.String{}
+	hasApproval := sets.String{}
```
sets.NewString()

grodrigues3 commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,62 +95,157 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
-		if !h.hasApproval(*file.Filename, approverSet, maxDepth) {
-			return
+	var copyOfFiles sets.String
+	for fn := range ownersPath {
```
uninitialized variable

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -17,20 +17,25 @@ limitations under the License.
+	"fmt"
 	"github.com/golang/glog"
-	githubapi "github.com/google/go-github/github"
+	goGithub "github.com/google/go-github/github"
```
as agreed ;-)

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,62 +96,165 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+		maxCovered := 0
+		var bestPerson string
+		for k, v := range approverCount {
+			if len(v) >= maxCovered && v.Intersection(copyOfFiles).Len() != 0 {
```
why isn't this:
```
v.Intersection(copyOfFiles).Len() > maxCovered
```

Why `>=` and why `len(v)` rather than intersection?

apelisse commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,62 +95,169 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+		glog.Error("Could not list the comments for PR%v", obj.Issue.Number)
+		return err
+	}
+	latestNotification := mungeComment.FilterComments(comments, notificationMatcher).GetLast()
```
Doesn't look like me, somebody must have forced me to do that during my sleep :-)

grodrigues3 commented on mungegithub/mungers/approval-handler.go:
```diff
@@ -90,62 +96,165 @@ func (h *ApprovalHandler) Munge(obj *github.MungeObject) {
+	i := 0
+	for path := range ownersMap {
+		sliceOfKeys[i] = path
+	}
```
keeping my implementation (with incr fix)

apelisse commented:
Looks good to me! :-)
/lgtm

I'll take care of squashing.

grodrigues3 commented:
#2076
